nav2_behavior_tree/include/nav2_behavior_tree/plugins/condition/is_path_valid_condition.hpp

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) 2021 Joshua Wallace
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NAV2_BEHAVIOR_TREE__PLUGINS__CONDITION__IS_PATH_VALID_CONDITION_HPP_
+#define NAV2_BEHAVIOR_TREE__PLUGINS__CONDITION__IS_PATH_VALID_CONDITION_HPP_
+
+#include <string>
+#include <memory>
+
+#include "rclcpp/rclcpp.hpp"
+#include "behaviortree_cpp_v3/condition_node.h"
+#include "geometry_msgs/msg/pose_stamped.hpp"
+#include "nav2_msgs/srv/is_path_valid.hpp"
+
+namespace nav2_behavior_tree
+{
+
+/**
+ * @brief A BT::ConditionNode that returns SUCCESS when the IsPathValid
+ * service returns true and FAILURE otherwise
+ */
+class IsPathValidCondition : public BT::ConditionNode
+{
+public:
+  /**
+   * @brief A constructor for nav2_behavior_tree::IsPathValidCondition
+   * @param condition_name Name for the XML tag for this node
+   * @param conf BT node configuration
+   */
+  IsPathValidCondition(
+    const std::string & condition_name,
+    const BT::NodeConfiguration & conf);
+
+  IsPathValidCondition() = delete;
+
+  /**
+   * @brief The main override required by a BT action
+   * @return BT::NodeStatus Status of tick execution
+   */
+  BT::NodeStatus tick() override;
+
+  /**
+   * @brief Creates list of BT ports
+   * @return BT::PortsList Containing node-specific ports
+   */
+  static BT::PortsList providedPorts()
+  {
+    return {
+      BT::InputPort<nav_msgs::msg::Path>("path", "Path to Check"),
+      BT::InputPort<std::chrono::milliseconds>("server_timeout")
```

Comment:
Hi,
Maybe there should be a default value set for the timeout (e.g.5-10ms as suggested in previous comments).
Here the default is 0ms, and when experimenting with this PR, it took me a little time to realize that IsPathValid was always failing because of that.

Reply:
Agreed, what's the default used elsewhere?

Reply:
The default is 20ms and it gets declared in the 

https://github.com/ros-planning/navigation2/blob/main/nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp on line 111. 

This got set correctly on my blackboard so I am not sure why @Kaetera is seeing it get set to 0ms. 

Do we want to add another parameter to nav2_params.yaml for service_timeout?

Reply:
No, you need to add a default to the InputPort, right now there's no default value in it so it would use the initialized value of 0. Here's an example of a input port with a default set https://github.com/ros-planning/navigation2/blob/main/nav2_behavior_tree/include/nav2_behavior_tree/plugins/action/back_up_action.hpp#L56

Reply:
So, it looks like std::to_string has no def for chrono types which makes sense. 

/opt/ros/rolling/include/behaviortree_cpp_v3/basic_types.h:138:26: error: no matching function for call to ‘to_string(std::chrono::duration<long int, std::ratio<1, 1000> >&)’
  138 |     return std::to_string(value);
  
  This prevents the default being set. I spent some time trying to figure it out yesterday did not have any luck.

Reply:
Got same problem, I ended to make this param an int to get the value as number of milliseconds, and then convert it later in an std::chrono duration (milliseconds or seconds, we should use the same unit as used in other BT nodes)

Reply:
That is a possible solution but quite fragile. It would depend on people correctly converting and would result in a lot of file changes. I think a better solution would be to just have the timeout in the xml. Maybe a note also in the code too?

Reply:
Actually when using a node with such param in the XML BTree, all params are entered as strings, so it does not really matters if it's cast to an int, float or chrono in the code. The user just has to know the unit expected, a good naming and documentation should be sufficient, like "server_timeout_ms". We already do rely on the documentation/naming for all the other nodes, e.g. to enter the `backup_dist` of the BackUp recovery node in meters, so I think it's ok to do the same for this timeout.
It involves only changing one or two lines in is_path_valid_condition.hpp and is_path_valid_condition.cpp (unless I missed some tests scripts that set this parameter).

Reply:
That's what we use in all other contexts, https://github.com/ros-planning/navigation2/blob/779e3b60b9237374a76ed42233b116786715c575/nav2_behavior_tree/include/nav2_behavior_tree/plugins/action/spin_action.hpp#L57 have some seconds / milliseconds and it can be converted to chrono, this isn't fragile (or if it is, then its just as fragile as the ROS parameter systems which also doesn't support Chrono types in the yaml; nothing does for IO parsing...).

Reply:
https://github.com/ros-planning/navigation2/blob/806029d0652c2618400f13fe5822f8ad671e1371/nav2_behavior_tree/include/nav2_behavior_tree/bt_service_node.hpp#L56-L58

Similarly, you could use the global service timeout set on the blackboard instead. Then have the input port be used **only if set** and so you don't need to specifically add a default. So you could keep this code the same without a default but then get the default from the  blackboard. And this code also does seem to show use of chrono types since we defined serialization/deserialization overrides for it in other header files we include.

Reply:
Okay, so I switched it to a int and tested. Everything looks good. 

So deserialization does not seem to be an issue only serialization. 

https://github.com/ros-planning/navigation2/blob/806029d0652c2618400f13fe5822f8ad671e1371/nav2_behavior_tree/include/nav2_behavior_tree/bt_service_node.hpp#L56-L58

Also, this is a little odd. It seems like we are setting server_timeout twice?

Reply:
Its looking from 2 sources but only uses the best value.

The blackboard has a generic timeout used by everything that wasn't explicitly set. Then we have a default-free version in BT nodes that use the timeout so they can be individually overrided. So what's happening is we use the global default from the blackboard **unless** the local server timeout port as a non-default value to set (which getInput handles). 

And its using Chrono types :-)

Reply:
@jwallace42 please use the same strategy here: https://github.com/ros-planning/navigation2/blob/806029d0652c2618400f13fe5822f8ad671e1371/nav2_behavior_tree/include/nav2_behavior_tree/bt_service_node.hpp#L56-L58 with the same variables and default to blackboard values.

Reply:
Done